ledger/trie/light_node.go

```diff
@@ -0,0 +1,316 @@
+	hash, err := hash.ToHash(ln.HashValue)
+	if err != nil {
+		return nil, fmt.Errorf("invalid hash in light node: %w", err)
+	}
+
+	if len(ln.Path) == 0 {
+		// Since it does not have a path, this node is a branch.
+		return &Branch{
+			lChild: nodes[ln.LIndex],
+			rChild: nodes[ln.RIndex],
+			height: ln.Height,
+			hash:   hash,
+			dirty:  false,
+		}, nil
+	}
+
+	path, err := ledger.ToPath(ln.Path)
+	if err != nil {
+		return nil, fmt.Errorf("invalid path in light node: %w", err)
+	}
+
+	if ln.Skip > 0 {
+		// Since it has a skip value, this node is an extension.
+		return &Extension{
+			lChild: nodes[ln.LIndex],
+			rChild: nodes[ln.RIndex],
+			height: ln.Height,
+			path:   path,
+			skip:   ln.Skip,
+			hash:   hash,
+			dirty:  false,
+		}, nil
+	}
+
+	// Since it has a path and has no skip value, this node is a leaf.
+	return &Leaf{
+		path:   path,
+		hash:   hash,
+		height: ln.Height,
+	}, nil
+}
+
+const (
+	legacyVersion   = uint16(0)
+	encodingVersion = uint16(1)
+)
+
+func EncodeLightNode(lightNode *LightNode, store dps.Store) []byte {
+
+	var payload *ledger.Payload
+	if len(lightNode.Path) > 0 && lightNode.Skip == 0 {
+		// Since the node has a path and no skip value, we know it is a leaf node.
+		h, err := hash.ToHash(lightNode.HashValue)
+		if err != nil {
+			panic(fmt.Errorf("fatal error: invalid hash in node: %w", err))
+		}
+
+		payload, err = store.Retrieve(h)
+		if err != nil {
+			panic(fmt.Errorf("fatal error: missing payload from store: %w", err))
+		}
+	}
+	encPayload := encoding.EncodePayload(payload)
+
+	length := 2 + 8 + 8 + 2 + 2 + len(lightNode.Path) + len(lightNode.HashValue) + len(encPayload)
+	buf := make([]byte, length)
+
+	buf = utils.AppendUint16(buf, encodingVersion)
+	buf = utils.AppendUint64(buf, lightNode.LIndex)
+	buf = utils.AppendUint64(buf, lightNode.RIndex)
+	buf = utils.AppendUint16(buf, lightNode.Height)
+	buf = utils.AppendUint16(buf, lightNode.Skip)
+	buf = utils.AppendShortData(buf, lightNode.Path)
+	buf = utils.AppendShortData(buf, lightNode.HashValue)
+	buf = utils.AppendLongData(buf, encPayload)
+
+	return buf
+}
+
+func DecodeLightNode(reader io.Reader, store dps.Store) (*LightNode, error) {
+	var buf [2]byte
+	read, err := io.ReadFull(reader, buf[:])
+	if err != nil {
+		return nil, fmt.Errorf("could not read light node encoding version: %w", err)
+	}
+	if read != len(buf) {
+		return nil, fmt.Errorf("not enough bytes read (got %d, expected %d)", read, len(buf))
+	}
+
+	version, _, err := utils.ReadUint16(buf[:])
+	if err != nil {
+		return nil, fmt.Errorf("could not read light node: %w", err)
+	}
+
+	switch version {
+	case encodingVersion:
+		return decodeLightNode(reader, store)
+	case legacyVersion:
+		return decodeLegacyNode(reader, store)
+	default:
+		return nil, fmt.Errorf("unsupported encoding version: expected %d got %d", encodingVersion, version)
+	}
+}
+
+func decodeLightNode(reader io.Reader, store dps.Store) (*LightNode, error) {
+	length := 8 + 8 + 2 + 2
```

Comment:
I can add a comment for it yes. There is one somewhere in this file that describes the length of each field, but you're right, it would be nice to have them everywhere we do this.